spec/mongo/operation/create/op_msg_spec.rb

```diff
@@ -0,0 +1,285 @@
+# frozen_string_literal: true
+# encoding: utf-8
+
+require 'spec_helper'
+
+describe Mongo::Operation::Create::OpMsg do
+
+  let(:write_concern) do
+    Mongo::WriteConcern.get(w: :majority)
+  end
+
+  let(:session) { nil }
+  let(:spec) do
+    { :selector       => { :create => authorized_collection.name },
+      :db_name       => authorized_collection.database.name,
+      :write_concern => write_concern,
+      :session       => session
+    }
+  end
+
+  let(:op) { described_class.new(spec) }
+
+  let(:connection) do
+    double('connection').tap do |connection|
+      allow(connection).to receive(:server).and_return(authorized_primary)
+      allow(connection).to receive(:features).and_return(authorized_primary.features)
+      allow(connection).to receive(:description).and_return(authorized_primary.description)
+      allow(connection).to receive(:cluster_time).and_return(authorized_primary.cluster_time)
+    end
+  end
+
+  describe '#initialize' do
+
+    context 'spec' do
+
+      it 'sets the spec' do
+        expect(op.spec).to eq(spec)
+      end
+    end
+  end
+
+  describe '#==' do
+
+    context 'spec' do
+
+      context 'when two ops have the same specs' do
+        let(:other) { described_class.new(spec) }
+
+        it 'returns true' do
+          expect(op).to eq(other)
+        end
+      end
+
+      context 'when two ops have different specs' do
+        let(:other_selector) do
+           { :create => "other_collection_name" }
+        end
+
+        let(:other_spec) do
+          { :selector       => other_selector,
+            :db_name       => authorized_collection.database.name,
+            :write_concern => write_concern,
+            :ordered       => true
+          }
+        end
+        let(:other) { described_class.new(other_spec) }
+
+        it 'returns false' do
+          expect(op).not_to eq(other)
+        end
+      end
+    end
+  end
+
+  describe '#selector' do
+    min_server_fcv '3.6'
+
+    it 'does not mutate user input' do
+      user_input = spec.dup
+      described_class.new(user_input).send(:selector, connection)
+      expect(user_input).to eq(spec)
```

Comment:
This does not catch deep mutation. We have https://github.com/dkubb/ice_nine attached to the test suite, so  I think what would work here is

```
user_input = IceNine.deep_freeze(spec.dup)
```

and if this is modified I expect an exception to be raised.

Reply:
Thanks, good point! Changed the test as suggested. Seems that we are fine, no exception raised.

Reply:
Good point, thank you! Updated the test.